Refactor to a single-phase process but keep the three-stage Containerfile

AlanCoding opened the pull request:
Slightly revised version of https://github.com/ansible/ansible-builder/pull/216

When you run `ansible-builder build`, you get a `Containerfile`. Before any of this refactor, that was a 3-phase process.

With #216, you get a 2-phase process where `ansible-galaxy collection install`, `ansible-builder introspect`, and `assemble` happen in the same stage.

This PR changes it so that we keep the `galaxy` and then `builder` stages. The `galaxy` stage uses the user-specified base image to do the galaxy install. The `builder` stage uses the new ansible-builder image to run both `ansible-builder introspect` and `assemble`.

### Pro arguments

 - By removing Ansible core from the `ansible/ansible-builder` image, we avoid situations where its version differs from the `ansible/ansible-runner` image. There are several known cases where the version of Ansible affects the behavior of the `ansible-galaxy` CLI.
 - This makes the new `ansible/ansible-builder` image smaller
 - If we, at some point, we combine ansible-builder and python-builder images, it would be a smaller change to do that

### Con arguments

 - This makes the `Containerfile` output for EEs more complex and harder to explain

shanemcd committed:
WIP

AlanCoding committed:
Add back --write-bindep type options as means of writing introspect output

AlanCoding committed:
WIP Initialize an ansible-builder image

AlanCoding committed:
Fix bug

AlanCoding committed:
Fix tagging issues with make image

AlanCoding committed:
Rename the build args

AlanCoding committed:
Combine introspect and assemble steps to get functionality

The ideal was that we could write multiple files for
  both requirements.txt and bindep.txt
  and then feed those into assemble and install-from-bindep

The current state of things is that /tmp/src/ only takes 1 primary
  requirements file for both of these,
  this brings us back to the need for --user-bindep and --user-pip
    where were previously implemented before other refactors happened

Big picture, that prior work was pushing things into the container
  later work started to push things back OUT of the container
  this work reverses that again, pushing things back INTO the container

AlanCoding committed:
Pull in missing items from the parent python builder image

This was missing the /output folder, but it needs to be a single copy
  if not, then it will copy over artifacts from the install of
  ansible-builder itself, which would be super meta, but wrong

bindep is added as a python dependency of ansible-builder
  yes, this is kind of a big deal
  for now, it just allows us to not worry about installing it
    inside of the ansible-builder image

AlanCoding committed:
Fix parser bug

AlanCoding committed:
Fix issue with missing directory and not being in /output

AlanCoding committed:
Break out ADD of build context from GalaxyInstallSteps

motivation: use case of user-provided requirements.txt was not working
  this fixes that

the introspect layer now makes use of files from the build context
  that creates a need to have these separated,
  but they were already fairly distinct in function

Add some more verbosity to introspect error cases
  this is a likely error case debugging this logic
  paths are inside of container
    so much more difficult for user to interactively poke around

AlanCoding committed:
Have tox make me an image so integration tests will run

AlanCoding committed:
Skip some stages conditionally to make no-op scenarios pass

AlanCoding committed:
Pre-build docker images for integration tests too

AlanCoding committed:
Remove test assertion no longer valid

AlanCoding committed:
WIP put in jobs for building and uploading ansible-builder image

AlanCoding committed:
Add a dockerignore file

AlanCoding committed:
Resolve a few TODO items

AlanCoding committed:
Change parents of ansible-builder image, back to 3-stage builds

This changes the parents of the ansible/ansible-builder image
  changes direct parent from ansible/ansible-runner to ansible/python-base
  continues to use python-builder in mostly the same way

The big takeaway is that ansible/ansible-builder will become a minimal image
  it will no longer have Ansible inside of it
  it will also no longer have ansible-runner

Removing Ansible from the parent image, this creates the need to have
  an initial "galaxy" stage in the EE Containerfile again
  so this implements that

AlanCoding committed:
whoops

AlanCoding commented:
The awx-ee built with this:

https://github.com/AlanCoding/awx-ee/compare/more_updates...AlanCoding:builder_facelift2?expand=1

AlanCoding committed:
Move build-arg to start of stage

AlanCoding commented on Containerfile:
```diff
@@ -0,0 +1,24 @@
+COPY --from=python_builder /usr/local/bin/get-extras-packages /usr/local/bin/get-extras-packages
+
+# building EEs require the install-from-bindep script, but not the rest of the /output folder
+COPY --from=python_builder /output/install-from-bindep /output/install-from-bindep
```
a little awkward, we could modify the above `&& rm -rf /output` to keep the file instead... but it's minor enough that the easier path might be preferable.

beeankha commented:
> The awx-ee built with this:
> 
> https://github.com/AlanCoding/awx-ee/compare/more_updates...AlanCoding:builder_facelift2?expand=1

@AlanCoding as discussed during the code walkthrough that you did with me, in the `awx-ee/Containerfile` you're linking to in the comment above, the line:
`RUN ansible-builder introspect --user-bindep=bindep.txt --write-bindep=/tmp/src/bindep.txt --write-pip=/tmp/src/requirements.txt`

...should most likely be:
`RUN ansible-builder --sanitize --user-bindep=bindep.txt --write-bindep=/tmp/src/bindep.txt --write-pip=/tmp/src/requirements.txt`

...due to the changes being made in this PR (`--sanitze` vs `introspect`).

beeankha reviewed (approved):
:+1:

AlanCoding committed:
Fix bug where --sanitize did not work with --write options

AlanCoding committed:
Always use the --sanitize flag for introspect in Containerfile

Also add a test to confirm that its working, add a package
  to the pip test case that is known to be in the ignore list

AlanCoding commented:
Fixed the noted `--sanitize` issue and added test coverage with the latest commits. The updated Containerfile for awx-ee now looks like this:

```Dockerfile
ARG EE_BASE_IMAGE=quay.io/ansible/ansible-runner:devel
ARG EE_BUILDER_IMAGE=quay.io/ansible/ansible-builder:latest

FROM $EE_BASE_IMAGE as galaxy
ARG ANSIBLE_GALAXY_CLI_COLLECTION_OPTS=
USER root

ADD _build /build
WORKDIR /build

RUN ansible-galaxy role install -r requirements.yml --roles-path /usr/share/ansible/roles
RUN ansible-galaxy collection install $ANSIBLE_GALAXY_CLI_COLLECTION_OPTS -r requirements.yml --collections-path /usr/share/ansible/collections

FROM $EE_BUILDER_IMAGE as builder

COPY --from=galaxy /usr/share/ansible /usr/share/ansible

ADD _build/bindep.txt bindep.txt
RUN ansible-builder introspect --sanitize --user-bindep=bindep.txt --write-bindep=/tmp/src/bindep.txt --write-pip=/tmp/src/requirements.txt
RUN assemble

FROM $EE_BASE_IMAGE
USER root

COPY --from=galaxy /usr/share/ansible /usr/share/ansible

COPY --from=builder /output/ /output/
RUN /output/install-from-bindep && rm -rf /output/wheels
RUN alternatives --set python /usr/bin/python3
COPY --from=quay.io/project-receptor/receptor:0.9.7 /usr/bin/receptor /usr/bin/receptor
RUN mkdir -p /var/run/receptor
ADD run.sh /run.sh
CMD /run.sh
USER 1000
RUN git lfs install
```

(updated the awx-ee branch too)

AlanCoding commented:
Base branch for that awx-ee demo has changed, so I'll link it again

https://github.com/ansible/awx-ee/compare/devel...AlanCoding:builder_facelift2?expand=1

AlanCoding commented:
Leaving this diagram here publicly:

![Core EE Build Flow - May_2021 (1)](https://user-images.githubusercontent.com/1385596/119673927-c68c5200-be09-11eb-81f7-043fa19283c2.png)
